corehq/apps/app_manager/id_strings.py

```diff
@@ -171,9 +171,21 @@ def detail(module, detail_type):
     return u"m{module.id}_{detail_type}".format(module=module, detail_type=detail_type)
 
 
-def menu(module):
+def menu_id(module):
     put_in_root = getattr(module, 'put_in_root', False)
-    return ROOT if put_in_root else u"m{module.id}".format(module=module)
+    if put_in_root:
+        # handle circular calls, if bad module workflow setup
+        return menu_id(module.root_module) if getattr(module, 'root_module', False) else ROOT
+    else:
+        return u"m{module.id}".format(module=module)
+
+
+def menu_root(module):
```

Comment:
I think it would be cleaner to return None and have the caller check for
that

On 3 March 2015 at 09:10, Sravan Reddy <notifications@github.com> wrote:

> In corehq/apps/app_manager/id_strings.py
> <https://github.com/dimagi/commcare-hq/pull/5716#discussion_r25666946>:
>
> >      put_in_root = getattr(module, 'put_in_root', False)
> > -    return ROOT if put_in_root else u"m{module.id}".format(module=module)
> > +    if put_in_root:
> > +        # handle circular calls, if bad module workflow setup
> > +        return menu_id(module.root_module) if getattr(module, 'root_module', False) else ROOT
> > +    else:
> > +        return u"m{module.id}".format(module=module)
> > +
> > +
> > +def menu_root(module):
>
> This could still return root instead of empty, and the caller function
> can decide not to set root="root" if this returns root?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5716/files#r25666946>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa